"[p]muteset makerole" can error out and fail if ratelimited

# Command bugs

#### Command name

`[p]muteset makerole`

#### What cog is this command from?

Mutes

#### What were you expecting to happen?

The command to successfully create a new role with overrides configured for a server's channels and categories

#### What actually happened?

It causes an exception, leaving a partially-configured role and posting an error in chat. The traceback ends with `discord.ext.commands.errors.CommandInvokeError: Command raised an exception: HTTPException: 429 Too Many Requests (error code: 0): You are being rate limited.`

#### How can we reproduce this issue?

1. Start with a server that has no mute role configured, and a considerably large number of channels.
2. Run `[p]muteset makerole [role]`
3. If Discord ratelimits Red while it's applying overrides, the command will error out leaving the role only partially configured.

My bad, does this work?:
```
redbot[50180]: [2020-10-26 04:09:45] [ERROR] red: Exception in command 'muteset makerole'
redbot[50180]: Traceback (most recent call last):
redbot[50180]:   File "/home/redbeta/.local/lib/python3.8/site-packages/discord/ext/commands/core.py", line 85, in wrapped
redbot[50180]:     ret = await coro(*args, **kwargs)
redbot[50180]:   File "/home/redbeta/.local/lib/python3.8/site-packages/redbot/cogs/mutes/mutes.py", line 697, in make_mute_role
redbot[50180]:     errors = await asyncio.gather(*tasks)
redbot[50180]:   File "/home/redbeta/.local/lib/python3.8/site-packages/redbot/cogs/mutes/mutes.py", line 729, in _set_mute_role_overwrites
redbot[50180]:     await channel.set_permissions(role, overwrite=overs, reason=_("Mute role setup"))
redbot[50180]:   File "/home/redbeta/.local/lib/python3.8/site-packages/discord/abc.py", line 658, in set_permissions
redbot[50180]:     await http.edit_channel_permissions(self.id, target.id, allow.value, deny.value, perm_type, reason=reason)
redbot[50180]:   File "/home/redbeta/.local/lib/python3.8/site-packages/discord/http.py", line 258, in request
redbot[50180]:     raise HTTPException(r, data)
redbot[50180]: discord.errors.HTTPException: 429 Too Many Requests (error code: 0): You are being rate limited.
```

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -1380,7 +1380,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
         else:
             results = []
             for channel in guild.channels:
-                result.append(await self.channel_unmute_user(guild, channel, author, user, reason))
+                results.append(await self.channel_unmute_user(guild, channel, author, user, reason))
             for task in results:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -1380,7 +1380,9 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
         else:
             results = []
             for channel in guild.channels:
-                results.append(await self.channel_unmute_user(guild, channel, author, user, reason))
+                results.append(
+                    await self.channel_unmute_user(guild, channel, author, user, reason)
+                )
             for task in results:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -322,13 +322,12 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
         self, member: discord.Member, guild: discord.Guild, channels: Dict[int, dict]
     ):
         """This is meant to handle multiple channels all being unmuted at once"""
-        tasks = []
+        results = []
         for channel, mute_data in channels.items():
             author = guild.get_member(mute_data["author"])
-            tasks.append(
-                self._auto_channel_unmute_user(guild.get_channel(channel), mute_data, False)
+            results.append(
+                await self._auto_channel_unmute_user(guild.get_channel(channel), mute_data, False)
             )
-        results = await asyncio.gather(*tasks)
         unmuted_channels = [guild.get_channel(c) for c in channels.keys()]
         for result in results:
             if not result:
@@ -697,10 +696,9 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             except discord.errors.Forbidden:
                 return await ctx.send(_("I could not create a muted role in this server."))
             errors = []
-            tasks = []
+
             for channel in ctx.guild.channels:
-                tasks.append(self._set_mute_role_overwrites(role, channel))
-            errors = await asyncio.gather(*tasks)
+                errors.append(await self._set_mute_role_overwrites(role, channel))
             if any(errors):
                 msg = _(
                     "I could not set overwrites for the following channels: {channels}"
@@ -1318,10 +1316,9 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             return ret
         else:
             perms_cache = {}
-            tasks = []
+            task_result = []
             for channel in guild.channels:
-                tasks.append(self.channel_mute_user(guild, channel, author, user, until, reason))
-            task_result = await asyncio.gather(*tasks)
+                task_result.append(await self.channel_mute_user(guild, channel, author, user, until, reason))
             for task in task_result:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))
@@ -1379,10 +1376,9 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             ret["success"] = True
             return ret
         else:
-            tasks = []
+            results = []
             for channel in guild.channels:
-                tasks.append(self.channel_unmute_user(guild, channel, author, user, reason))
-            results = await asyncio.gather(*tasks)
+                result.append(await self.channel_unmute_user(guild, channel, author, user, reason))
             for task in results:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -679,6 +679,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
     @muteset.command(name="makerole")
     @checks.admin_or_permissions(manage_roles=True)
     @commands.bot_has_guild_permissions(manage_roles=True)
+    @commands.max_concurrency(1, commands.BucketType.guild)
     async def make_mute_role(self, ctx: commands.Context, *, name: str):
         """Create a Muted role.
 
@@ -688,6 +689,15 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
         If you already have a muted role created on the server use
         `[p]muteset role ROLE_NAME_HERE`
         """
+        if await self.config.guild(ctx.guild).mute_role():
+            command = f"`{ctx.clean_prefix}muteset role`"
+            return await ctx.send(
+                _(
+                    "There is already a mute role setup in this server. "
+                    "Please remove it with {command} before trying to "
+                    "create a new one."
+                ).format(command=command)
+            )
         async with ctx.typing():
             perms = discord.Permissions()
             perms.update(send_messages=False, speak=False, add_reactions=False)
@@ -695,6 +705,8 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
                 role = await ctx.guild.create_role(
                     name=name, permissions=perms, reason=_("Mute role setup")
                 )
+                await self.config.guild(ctx.guild).mute_role.set(role.id)
+                # save the role early incase of issue later
             except discord.errors.Forbidden:
                 return await ctx.send(_("I could not create a muted role in this server."))
             errors = []
@@ -708,7 +720,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
                 ).format(channels=humanize_list([i for i in errors if i]))
                 for page in pagify(msg, delims=[" "]):
                     await ctx.send(page)
-            await self.config.guild(ctx.guild).mute_role.set(role.id)
+
             await ctx.send(_("Mute role set to {role}").format(role=role.name))
         if not await self.config.guild(ctx.guild).notification_channel():
             command_1 = f"`{ctx.clean_prefix}muteset notification`"

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -12,7 +12,7 @@ from .voicemutes import VoiceMutes
 
 from redbot.core.bot import Red
 from redbot.core import commands, checks, i18n, modlog, Config
-from redbot.core.utils import bound_gather
+from redbot.core.utils import bounded_gather
 from redbot.core.utils.chat_formatting import humanize_timedelta, humanize_list, pagify
 from redbot.core.utils.mod import get_audit_reason
 from redbot.core.utils.menus import start_adding_reactions
@@ -329,7 +329,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             tasks.append(
                 self._auto_channel_unmute_user(guild.get_channel(channel), mute_data, False)
             )
-        results = await bound_gather(*tasks)
+        results = await bounded_gather(*tasks)
         unmuted_channels = [guild.get_channel(c) for c in channels.keys()]
         for result in results:
             if not result:
@@ -701,7 +701,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             tasks = []
             for channel in ctx.guild.channels:
                 tasks.append(self._set_mute_role_overwrites(role, channel))
-            errors = await bound_gather(*tasks)
+            errors = await bounded_gather(*tasks)
             if any(errors):
                 msg = _(
                     "I could not set overwrites for the following channels: {channels}"
@@ -1322,7 +1322,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             tasks = []
             for channel in guild.channels:
                 tasks.append(self.channel_mute_user(guild, channel, author, user, until, reason))
-            task_result = await bound_gather(*tasks)
+            task_result = await bounded_gather(*tasks)
             for task in task_result:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))
@@ -1383,7 +1383,7 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             tasks = []
             for channel in guild.channels:
                 tasks.append(self.channel_unmute_user(guild, channel, author, user, reason))
-            results = await bound_gather(*tasks)
+            results = await bounded_gather(*tasks)
             for task in results:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -12,6 +12,7 @@ from .voicemutes import VoiceMutes
 
 from redbot.core.bot import Red
 from redbot.core import commands, checks, i18n, modlog, Config
+from redbot.core.utils import bound_gather
 from redbot.core.utils.chat_formatting import humanize_timedelta, humanize_list, pagify
 from redbot.core.utils.mod import get_audit_reason
 from redbot.core.utils.menus import start_adding_reactions
@@ -322,12 +323,13 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
         self, member: discord.Member, guild: discord.Guild, channels: Dict[int, dict]
     ):
         """This is meant to handle multiple channels all being unmuted at once"""
-        results = []
+        tasks = []
         for channel, mute_data in channels.items():
             author = guild.get_member(mute_data["author"])
-            results.append(
-                await self._auto_channel_unmute_user(guild.get_channel(channel), mute_data, False)
+            tasks.append(
+                self._auto_channel_unmute_user(guild.get_channel(channel), mute_data, False)
             )
+        results = await bound_gather(*tasks)
         unmuted_channels = [guild.get_channel(c) for c in channels.keys()]
         for result in results:
             if not result:
@@ -696,9 +698,10 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             except discord.errors.Forbidden:
                 return await ctx.send(_("I could not create a muted role in this server."))
             errors = []
-
+            tasks = []
             for channel in ctx.guild.channels:
-                errors.append(await self._set_mute_role_overwrites(role, channel))
+                tasks.append(self._set_mute_role_overwrites(role, channel))
+            errors = await bound_gather(*tasks)
             if any(errors):
                 msg = _(
                     "I could not set overwrites for the following channels: {channels}"
@@ -1316,11 +1319,10 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             return ret
         else:
             perms_cache = {}
-            task_result = []
+            tasks = []
             for channel in guild.channels:
-                task_result.append(
-                    await self.channel_mute_user(guild, channel, author, user, until, reason)
-                )
+                tasks.append(self.channel_mute_user(guild, channel, author, user, until, reason))
+            task_result = await bound_gather(*tasks)
             for task in task_result:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))
@@ -1378,11 +1380,10 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             ret["success"] = True
             return ret
         else:
-            results = []
+            tasks = []
             for channel in guild.channels:
-                results.append(
-                    await self.channel_unmute_user(guild, channel, author, user, reason)
-                )
+                tasks.append(self.channel_unmute_user(guild, channel, author, user, reason))
+            results = await bound_gather(*tasks)
             for task in results:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

--- a/redbot/cogs/mutes/mutes.py
+++ b/redbot/cogs/mutes/mutes.py
@@ -1318,7 +1318,9 @@ class Mutes(VoiceMutes, commands.Cog, metaclass=CompositeMetaClass):
             perms_cache = {}
             task_result = []
             for channel in guild.channels:
-                task_result.append(await self.channel_mute_user(guild, channel, author, user, until, reason))
+                task_result.append(
+                    await self.channel_mute_user(guild, channel, author, user, until, reason)
+                )
             for task in task_result:
                 if not task["success"]:
                     ret["channels"].append((task["channel"], task["reason"]))

